docs/GraphOperationAPIs.md

```diff
@@ -0,0 +1,739 @@
+  static [Operation]Plan [Algo1](
+      unsigned arg = kDefault[arg], [...]) {
+    return {kCPU, k[Algo1], arg, [...]};
+  }
+  [...]
+};
+```
+
+### Operation Function
+
+The operation function takes a graph, any functional arguments, and a plan.
+It should return `katana::Results<void>`.
+If there are graph wide results (such as, the triangle count), the function may replace `void` with an appropriate value.
+
+The operation should not modify the graph except to add the output properties and to add/remove ephemeral properties used during the algorithm.
+If the algorithm must change the graph (for instance, to sort nodes or edges), it should copy the graph before making such changes and then copy any output property back to the original graph (with appropriate transformation) before it returns.
+
+```c++
+KATANA_EXPORT Result<void> [Operation](
+    PropertyFileGraph* pfg, [... functional arguments ...], [Operation]Plan plan = {});
+```
+
+### Validity checking function
+
+The `AssertValid` function should perform a fast check to see if the output of the operation is as expected.
+It will generally have false positives (cases where it could return no error when the data is in fact wrong).
+
+It generally takes the same arguments as the operation except for the plan.
+It can omit some functional arguments if they can be inferred from the data being checked.
+It returns `katana::ErrorCode::AssertionFailed` if the output is not as expected.
+
+```c++
+KATANA_EXPORT Result<void> [Operation]AssertValid(
+    PropertyFileGraph* pfg, [... functional arguments ...]);
+```
+
+This function can be omitted if there is no good way to check the results of the algorithm quickly.
+
+### Statistics
+
+The Statistics struct analyzes the output of the operation and produces useful statistics about the graph as a whole.
+The struct can `Print` itself (by default to `stdout` for debugging).
+The values are accessible directly as fields.
+Useful values computed from the fields can be included using small computed accessor methods.
+
+The struct is created using a `Compute` static method on the struct.
+It generally takes the same arguments as the operation except for the plan.
+It can omit some functional arguments if they can be inferred from the data being checked.
+
+```c++
+struct KATANA_EXPORT [Operation]Statistics {
+  [... statistics ...]
+
+  [... computed statistics methods ...]
+
+  /// Print the statistics in a human readable form.
+  void Print(std::ostream& os = std::cout) const;
+
+  static katana::Result<[Operation]Statistics> Compute(
+      PropertyFileGraph* pfg, [... functional arguments ...]);
+};
+```
+
+### Documentation
+
+The following elements should be documented, but you can document other things of course.
+
+1. Every constructor and static method constructor on the Plan.
+2. The operation function.
+3. Every member of the Statistics struct.
+
+### Complete Example: SSSP
+
+```c++
+#ifndef KATANA_LIBGALOIS_KATANA_ANALYTICS_SSSP_SSSP_H_
```

Comment:
Do we need the complete example here? Can we just have a pointer to the file?

Reply:
Yea. You are probably right. I'll change it.